agent/src/main/java/org/openjdk/jmc/agent/SapAgent.java

```diff
@@ -0,0 +1,345 @@
+
+	private static final String CONFIGS_PATH = "org/openjdk/jmc/agent/sap/";
+	private static final String CONVERTERS_PREFIX = "org.openjdk.jmc.agent.sap.boot.converters.";
+	private static Logger logger = Logger.getLogger(SapAgent.class.getName());
+	private static Instrumentation instr;
+	private static boolean addedBootJar = false;
+	private static HashMap<String, String> seenCommands = new HashMap<>();
+
+	public static void premain(String agentArguments, Instrumentation instrumentation) throws Exception {
+		agentmain(agentArguments, instrumentation);
+	}
+
+	public static void agentmain(String agentArguments, Instrumentation instrumentation) throws Exception {
+		ModuleUtils.openUnsafePackage(instrumentation);
+		instr = instrumentation;
+
+		if (agentArguments == null || agentArguments.trim().length() == 0) {
+			initializeAgent(null, instrumentation);
+		} else {
+			if (agentArguments.equals("help")) {
+				ensureBootJarAdded();
+				Commands.printAllCommands();
+				System.exit(0);
+			}
+
+			if (agentArguments.startsWith("dump=")) {
+				ensureBootJarAdded();
+				Dumps.performDump(agentArguments.substring(5));
+				return;
+			}
+
+			try (InputStream stream = new ByteArrayInputStream(getXmlConfig(agentArguments))) {
+				initializeAgent(stream, instrumentation);
+			} catch (XMLStreamException | IOException | XMLValidationException e) {
+				logger.log(Level.SEVERE, "Failed to read jfr probe definitions from " + agentArguments, e);
+			}
+		}
+	}
+
+	private static void addBootJarIfNeeded(TransformRegistry registry) throws IOException {
+		// Check if we need converters from the boot jar.
+		boolean needsBootJar = false;
+
+		outer: for (String className : registry.getClassNames()) {
+			for (TransformDescriptor descriptor : registry.getTransformData(className)) {
+				JFRTransformDescriptor impl = (JFRTransformDescriptor) descriptor;
+
+				for (Field field : impl.getFields()) {
+					if (field.hasConverter() && field.getConverterDefinition().startsWith(CONVERTERS_PREFIX)) {
+						needsBootJar = true;
+						break outer;
+					}
+				}
+
+				for (Parameter param : impl.getParameters()) {
+					if (param.hasConverter() && param.getConverterDefinition().startsWith(CONVERTERS_PREFIX)) {
+						needsBootJar = true;
+						break outer;
+					}
+				}
+			}
+		}
+
+		if (needsBootJar) {
+			ensureBootJarAdded();
+		}
+	}
+
+	private static void ensureBootJarAdded() throws IOException {
+		if (addedBootJar) {
+			return;
+		}
+
+		ClassLoader cl = SapAgent.class.getClassLoader();
+
+		// Find out where the agent jar is, since the boot jar should live
+		// there too.
+		URL url = cl.getResource(SapAgent.class.getName().replace('.', '/') + ".class");
+		String file = url.getFile();
+
+		if (!file.startsWith("file:/")) {
+			throw new IOException("Could not determine agent jar from " + file);
+		}
+
+		System.err.println("Agent base file: " + file);
+
+		String os = AccessController.doPrivileged(new PrivilegedAction<String>() {
```

Comment:
Could probably be simplified to ?
`String os = AccessController.doPrivileged(() -> System.getProperty("os.name"));`

Reply:
With a proper type cast this worked.